### DISCUSSION TITLE
can ThanosRuler discover a HA Alertmanager cluster without load balancing?

### DISCUSSION THREAD

**Discussion by parkedwards** [prometheus-operator/prometheus-operator#4617] [Q&A] [answered]
hey there,

just a quick question on this - so I know that the prometheus-operator supports Prometheus -> HA Alertmanager via the prometheus CRD ✅ 

https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/high-availability.md#alertmanager

I'm curious to know if the ThanosRuler CRD supports a similar type of HA service discovery of Alertmanager, without load balancing via the Alertmanager Service resource.  The [Prometheus docs advise against doing this specifically](https://prometheus.io/docs/alerting/latest/alertmanager/#high-availability), and we're seeing some strange behavior when attempting ^

I've noticed that when we have a ThanosRuler + HA Alertmanager (> 1 replica), and we send alert events via the Service DNS, only one Alertmanager node receives the alerting events, and the other just hangs out

```
# alertmanager-node-1
level=debug ts=2022-03-04T18:34:12.707Z caller=dispatch.go:165 component=dispatcher msg="Received alert" alert=FooBarMissingMetrics[d498601][active]
level=debug ts=2022-03-04T18:34:12.707Z caller=dispatch.go:165 component=dispatcher msg="Received alert" alert=FooBarMissingMetrics[d498601][active]
level=debug ts=2022-03-04T18:34:12.712Z caller=dispatch.go:165 component=dispatcher msg="Received alert" alert=FooBarMissingMetrics[d498601][active]

# alertmanager-node-2
level=debug ts=2022-03-04T18:36:47.850Z caller=cluster.go:329 component=cluster memberlist="2022/03/04 18:36:47 [DEBUG] memberlist: Initiating push/pull sync with: 01FXB21T0GT3JJZBBSG391SDRW 172.18.21.207:9094\n"
level=debug ts=2022-03-04T18:37:30.821Z caller=cluster.go:329 component=cluster memberlist="2022/03/04 18:37:30 [DEBUG] memberlist: Stream connection from=172.18.21.207:52964\n"
level=debug ts=2022-03-04T18:37:47.852Z caller=cluster.go:329 component=cluster memberlist="2022/03/04 18:37:47 [DEBUG] memberlist: Initiating push/pull sync with: 01FXB21T0GT3JJZBBSG391SDRW 172.18.21.207:9094\n"
```

**Comment by simonpasquier** [comment, marked as answer]
You can use DNSSRV records to send alerts to all Ruler instances instead of just one random.
The recommendation is to use `spec.alertmanagersConfig` to pass the Alertmanager configuration to the Thanos ruler (see https://thanos.io/v0.25/components/rule.md/#alertmanager):

```yaml
alertmanagers:
- api_version: v2
  static_configs:
  - dnssrv+_web._tcp.alertmanager-operated.monitoring.svc
```

**Reply by parkedwards** [reply]
@simonpasquier you rock. that was it - thank you!
